refactor(juno): integrate common interface with rawDB as trieDB engine

MaksymMalicki opened the pull request:
1. This PR should be reviewed after #3360  
2. It integrates the common interfaces across the project, including in utilities and tests. This PR uses `rawDB` - direct pebbleDB reads/writes - as a trieDB engine.
3. Some adapting had to be also done for the prove logic inside of `storage.go` and `storage_test.go` to support both `trie` and `trie2` packages.
4. The experimental --new-state boolean CLI flag is now fully integrated. It can be used via the command line and in the `dbcmd` `Revert` utility.
5. A new command was added to the `Makefile`, to run the unit tests strictly with the new state and trie implementations in the future CI/CD:
```
test-new-state: clean-testcache rustdeps ## Run tests with new state
	JUNO_NEW_STATE=true go test $(GO_TAGS) -v ./...
```

infrmtcs commented on blockchain/blockchain_test.go:
```diff
@@ -724,3 +724,55 @@ func TestSubscribeL1Head(t *testing.T) {
+	return suList, blocks, nil
+}
+
+func BenchmarkBlockchainStore(b *testing.B) {
```
We should write this in a separate file, because this file is already quite long.

MaksymMalicki replied:
actually removed this benchmark, its not used anymore and not very precise

infrmtcs commented on blockchain/blockchain_test.go:
```diff
@@ -724,3 +724,55 @@ func TestSubscribeL1Head(t *testing.T) {
+	blocks := make([]*core.Block, samples)
+	for i := range samples {
+		fmt.Println("fetching", i)
+		su, err := gw.StateUpdate(context.Background(), uint64(i))
```
Use test context instead

infrmtcs commented on blockchain/blockchain_test.go:
```diff
@@ -724,3 +724,55 @@ func TestSubscribeL1Head(t *testing.T) {
+				require.NoError(b, chain.Store(blocks[j], &emptyCommitments, stateUpdates[j], nil))
+			}
+		}
+	})
```
We should write a for loop for `useNewState`. It's not only about less code, but also about guarantee that the structure of the 2 tests are the same.

infrmtcs commented on blockchain/blockchain_test.go:
```diff
@@ -724,3 +724,55 @@ func TestSubscribeL1Head(t *testing.T) {
+
+	suList := make([]*core.StateUpdate, samples)
+	blocks := make([]*core.Block, samples)
+	for i := range samples {
```
Setup time can take time. If it's the case, we can consider using `pool` or `iter` to query concurrently with rate limit.

infrmtcs commented on blockchain/blockchain.go:
```diff
@@ -703,3 +967,20 @@ func (b *Blockchain) StoreGenesis(
+			return err
+		}
+
+		return b.trieDB.Journal(stateUpdate.NewRoot)
```
Question: What happens if synchronizer writes a new block in the middle of this process?

MaksymMalicki commented on blockchain/blockchain_test.go:
```diff
@@ -724,3 +724,55 @@ func TestSubscribeL1Head(t *testing.T) {
+			return nil, nil, err
+		}
+		suList[i] = su
+		block, err := gw.BlockByNumber(context.Background(), uint64(i))
```
removed entirely

MaksymMalicki committed:
squash

fix unit tests

cleanups

renames

unit tests

interface modifications

linter

linter

linter

MaksymMalicki committed:
squash

cleanups

Integrate common interfaces into rawdb

fixes after rebase

minor linter changes

clean-up blockchain.go, fix linter

minor cleanups

state atomicity

remove node ordering

fixes after rebase

MaksymMalicki committed:
squash

fix unit tests

cleanups

renames

unit tests

interface modifications

linter

linter

linter

MaksymMalicki committed:
squash

fix unit tests

cleanups

Integrate common interfaces into rawdb

fixes after rebase

minor linter changes

clean-up blockchain.go, fix linter

unit tests

linter

minor cleanups

linter

linter

fixes after rebase

MaksymMalicki committed:
squash

cleanups

Integrate common interfaces into rawdb

fixes after rebase

minor linter changes

clean-up blockchain.go, fix linter

minor cleanups

clean cache

fix unit tests

fixes after rebase

MaksymMalicki committed:
squash

cleanups

Integrate common interfaces into rawdb

fixes after rebase

minor linter changes

clean-up blockchain.go, fix linter

minor cleanups

state atomicity

remove node ordering

fixes after rebase

EgeCaner commented on core/common_state.go:
```diff
@@ -17,11 +17,11 @@ type CommonState interface {
 
-	Update(
-		blockNum uint64,
+	Update(blockNum uint64,
```
```suggestion
	Update(
	     blockNum uint64,
```

MaksymMalicki committed:
changes to the atomicity

MaksymMalicki committed:
changes to the atomicity

MaksymMalicki committed:
code style

MaksymMalicki committed:
code style

MaksymMalicki committed:
fixes after rebase

MaksymMalicki committed:
unit tests

MaksymMalicki committed:
unit tests

MaksymMalicki committed:
chore: remove last staticcheck, organise mocks

MaksymMalicki committed:
chore: address first batch of comments on the PR

MaksymMalicki committed:
chore: fixes after rebase

MaksymMalicki committed:
chore: address the PR comments, use new felt constructors

MaksymMalicki committed:
test: fix unit tests

rodrodros committed:
fix(core/state): compilation errors

MaksymMalicki committed:
chore: address PR comments, fix unit tests

MaksymMalicki committed:
test: increase codecov

MaksymMalicki committed:
test: fix unit tests after rebase

MaksymMalicki committed:
chore: code cleanups

MaksymMalicki committed:
feat: add a state version guard

MaksymMalicki committed:
test: minor unit tests fixes

MaksymMalicki committed:
chore: adjustments after rebase

MaksymMalicki committed:
chore: clean-up tests, comments

MaksymMalicki commented on blockchain/blockchain_test.go:
```diff
@@ -243,17 +245,13 @@ func TestStore(t *testing.T) {
 		require.NoError(t, err)
 		assert.Equal(t, block0, headBlock)
 
-		root, err := chain.StateCommitment()
-		require.NoError(t, err)
-		assert.Equal(t, stateUpdate0.NewRoot, &root)
```
Actualy added a test helper function for this

MaksymMalicki committed:
fix: code cleanups, state backward compatibility

MaksymMalicki commented on blockchain/statebackend/deprecated.go:
```diff
@@ -14,23 +12,6 @@ type deprecatedStateBackend struct {
 	baseState
 }
 
-func (b *deprecatedStateBackend) StateCommitment() (felt.Felt, error) {
```
This method for bot `stateBackend` and `deprecatedStateBacked` was only used in the tests. If needed, can be added in the future, for now I removed it and swapped with a test helper called `chainStateCommitment` in the `blockchain_test.go`

MaksymMalicki commented on cmd/juno/dbcmd.go:
```diff
@@ -162,14 +178,27 @@ func dbRevert(cmd *cobra.Command, args []string) error {
 	}
 	defer database.Close()
 
+	if err = blockchain.ValidateStateVersion(database, newState); err != nil {
```
validation required before reverting

MaksymMalicki commented on core/receipt.go:
```diff
@@ -71,15 +71,15 @@ func messagesSentHash(messages []*L2ToL1Message) felt.Felt {
 
 type (
-	onTempTrieFunc     func(uint8, func(*trie.Trie) error) error
+	onTempTrieFunc     func(uint8, func(*trie2.Trie) error) error
```
those are static functions, used solely for the new Hash calculation. `trie2` is more performant for the Hashes, so we can swap those

MaksymMalicki committed:
chore: fixes after the rebase

MaksymMalicki committed:
fix: update rpcv10 storage to accept old and new trie

MaksymMalicki commented on core/state/history.go:
```diff
@@ -76,9 +76,10 @@ func (s *stateHistory) checkDeployed(addr *felt.Felt) error {
+		// TODO: return a dedicated error like ErrContractNotFound
+		// once deprecated state is removed. Using db.ErrKeyNotFound
+		// for backward compatibility.
+		return db.ErrKeyNotFound
```
I've decided to make the state fully backward compatible for now. Returning different errors for the state history lookups would force us to do a lot of `oldStateError || newStateError` checks. The errors can be changed to more descriptive ones after the old state is fully deprecated. Hence this TODO

MaksymMalicki committed:
chore: self review

MaksymMalicki commented on core/state/state.go:
```diff
@@ -275,7 +274,11 @@ func (s *State) Update(
 	}
 
-	return s.flush(blockNum, &stateUpdate, dirtyClasses, true)
+	// TODO: Remove this guard once the State and StateReader are split
```
Handled in the #3563

MaksymMalicki commented on vm/vm_test.go:
```diff
@@ -39,6 +47,7 @@ func TestCallDeprecatedCairo(t *testing.T) {
 	}, map[felt.Felt]core.ClassDefinition{
 		*classHash: simpleClass,
 	}, false))
+	require.NoError(t, batch.Write())
```
The new state can't read its own writes - database is used for reads. The changes applied to the batch by `state.Update` need to be applied to the DB.

MaksymMalicki committed:
chore: fixes after rebase

MaksymMalicki committed:
chore: fixes after rebase

MaksymMalicki committed:
refactor: remove state version logic

MaksymMalicki committed:
chore: cleanups

EgeCaner commented on node/node.go:
```diff
@@ -206,6 +208,11 @@ func New(cfg *Config, version string, logLevel *log.Level) (*Node, error) {
+		opts = append(opts, blockchain.WithNewState(
+			cfg.NewState,
+		))
+	}
```
Maybe we can add option regardless since option receives bool, following would provide the same outcome 
```suggestion
		opts = append(opts, blockchain.WithNewState(
			cfg.NewState,
		))
```

Let me know what do you think

MaksymMalicki replied:
Agreed

EgeCaner commented on node/node.go:
```suggestion
	opts := make([]blockchain.Option, 0, 2)
```

EgeCaner commented on vm/vm_test.go:
```diff
@@ -387,3 +424,24 @@ func TestSetVersionedConstants(t *testing.T) {
+	if err != nil {
+		return nil, err
+	}
+	return state, nil
```
```suggestion
	return state.New(stateRoot, stateDB, batch)
```

EgeCaner commented on rpc/v9/storage_test.go:
```diff
@@ -418,21 +474,22 @@ func TestStorageProof(t *testing.T) {
 		mockReader.EXPECT().BlockHeaderByNumber(blockNumber).
 			Return(headBlock.Header, nil)
 		mockState.EXPECT().ContractClassHash(noSuchKey).Return(felt.Zero, db.ErrKeyNotFound).Times(1)
+		mockState.EXPECT().ContractNonce(noSuchKey).Return(felt.Zero, db.ErrKeyNotFound).Times(0)
```
Why times 0?

MaksymMalicki replied:
could be removed, old regression

MaksymMalicki committed:
chore: address comments

MaksymMalicki committed:
refactor: revert `trie2` back to `trie` for receipt and tx

MaksymMalicki committed:
ci: include new state coverage in the CI
